### ISSUE TITLE
bound param names getting overwritten during set_parent_gremlin if values match

### ISSUE THREAD

**Issue by robmcdan** [emehrkay/gremlinpy#2]
When setting the parent gremlin of an anonymous traversal, the name key of of bound params is overwritten by the parent key if the values match.
### Repro

Notice that the key "GPY_PARAM_72e56_1" from the anonymous traversal still exists in the method parameter but it's key has been overwritten by the parent key "GPY_PARAM_04dc4_1" in the bound param dictionary -- as a result, one of the bound param keys is missing.

```
g = Gremlin()
a = g.V().AS("a").out("b").AS("c").unbound("where", __.AS("c").out("e"))
print(str(a))
print(a.bound_params)
>> g.V().as(GPY_PARAM_952de_1).out(GPY_PARAM_20818_2).as(GPY_PARAM_04dc4_1).where(__.as(GPY_PARAM_72e56_1).out(GPY_PARAM_0c59e_5))
>> {'GPY_PARAM_20818_2': 'b', 'GPY_PARAM_04dc4_1': 'c', 'GPY_PARAM_952de_1': 'a', 'GPY_PARAM_0c59e_5': 'e'}

```
### Cause

In Gremlin.bind_param, the key name is thrown away if its value already exists in the dictionary. When bind_param is called by set_parent_gremlin, this has the effect of leaving a key in the child gremlin which is not inherited by the parent.

I'm not sure if there is a reason to avoid duplicate values in the bound_params dict, or if this is an unintended consequence.

It looks like it could be fixed like this, inside of bind_param():

```
if value in self.stack_bound_params.values():
     for n, v in self.bound_params.items():
          if v == value and name is None:  # only borrow the parent name if a name hasn't already been set
               name = n
               break
```
### Workaround

You can avoid this by using nested instances and manually binding params.

```
g = Gremlin()
g1 = Gremlin("__")
b1 = g1.bind_param("c", "key")
b = g.bind_param("c", "key")
a = g.V().AS("a").out("b").AS(b[0]).where(g1.AS(b1[0]).out("e"))
print(str(a))
print(a.bound_params)
>> g.V().as(GPY_PARAM_5f1c4_2).out(GPY_PARAM_6a8f5_5).as(key).where(__.as(key).out(GPY_PARAM_b8840_7))
>> {'GPY_PARAM_6a8f5_5': 'b', 'GPY_PARAM_5f1c4_2': 'a', 'GPY_PARAM_b8840_7': 'e', 'key': 'c'}

```

**Comment by emehrkay** [comment]
Cool. Thank you for the write up. You are correct, we want to reuse already defined bound keys as this gives the Gremlin server less scripts to memoize during parsing and determine which to reuse.

I was able to reproduce this error and implement your fix. You didn't do a pull request though, so I just wrote it into the source and added a few tests. 

This was a good catch, I manually bind params using `Param(name, value)` (as should you) so often that this went unnoticed/untested. 

Thanks again,
Mark
